### PR TITLE
[#3] Setup CD pipeline (staging) - Final

### LINKED ISSUE #3: Set up CD pipeline for staging releases
## Why

When starting a new project, the CD pipeline should be setup properly for automatically testing to validate if changes to the codebase are correct and stable for immediate autonomous deployment to a staging environment.

## Acceptance Criteria
- [ ] Setup the CD pipeline for releasing the staging app using a preferred platform (Bitrise, CircleCI, Github Action, etc.) and release it on TestFlight or Firebase Distribution.
- [ ] Setup the triggers for every new push to develop branch.

### PR DESCRIPTION
**Pull request by markgravity**
Resolved #3 

## What happened

When starting a new project, the CD pipeline should be set up properly for automatically testing to validate if changes to the codebase are correct and stable for immediate autonomous deployment to a staging environment.

## Insight

- Add Fastlane to support sync signing information and upload to Firebase
- Add build & upload a staging build to Firebase using Github Action

## Proof Of Work

CI Result
Screen Shot 2021-08-20 at 16 29 25

Firebase
Screen Shot 2021-08-20 at 16 33 41

### CODE REVIEW

File: .github/workflows/deploy_staging.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    - name: Match Ad-hoc
+      run: bundle exec fastlane sync_adhoc_signing_ci
+      env:
+        MATCH_PASSWORD: ${{ secrets.MATCH_PASS }}
+
+    - name: Build App and Distribute to Firebase
+      run: bundle exec fastlane build_and_upload_staging_app
+      env:
+        FIREBASE_TOKEN: ${{ secrets.FIREBASE_TOKEN }}
```
Comment (minhnimble) [inline .github/workflows/deploy_staging.yml:78]: why don't we distribute to TestFlight? we can use a single app distribution place, no need to distribute it in many places 🙏

**Comment by minhnimble** [comment]
Need to fix conflicts for this branch 🛠🦀

**Comment by minhnimble** [comment]
Need to fix conflicts on this PR 🦀🛠

File: fastlane/Constants/Constants.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
+# frozen_string_literal: true
+
+class Constants
+  #################
+  #### PROJECT ####
+  #################
+
+    #Project path
```
Comment (minhnimble) [inline fastlane/Constants/Constants.rb:8]: the indentation and spacing here is wrong, please update it 🙏
```suggestion
# Project path
```

File: fastlane/Constants/Constants.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    './NimbleMedium.xcodeproj'
+  end
+
+  def self.BUNDLE_ID_STAGING
+    'co.nimblehq.nimble-medium-stag'
+  end
+
+  def self.BUNDLE_ID_PRODUCTION
+    'co.nimblehq.nimble-medium'
+  end
```
Comment (minhnimble) [inline fastlane/Constants/Constants.rb:19]: please add comments for these strings as well

**Review by minhnimble — APPROVED** [review]

**Review by suho — APPROVED** [review]

**Pull request merged**